controllers/consumerCouponController.js

```diff
@@ -0,0 +1,144 @@
+const { generateBase64QRCode } = require("../helpers/generateBase64QRCode");
+
+const ConsumerCoupon = require("../models/ConsumerCoupon");
+
+const getAllConsumerCoupons = async (req, res) => {
+  try {
+    const consumerId = req.consumerId;
+
+    if (!consumerId) {
+      return res.status(404).json({ error: "Consumer not found" });
+    }
+
+    // Find all consumer coupons for the consumer and populate consumer and coupon fields
+    const consumerCoupons = await ConsumerCoupon.find({ consumer: consumerId })
+      .populate({
+        path: "consumer",
+        populate: { path: "user", select: "-password" },
+      })
+      // .populate({
+      //   path: "coupon",
+      //   populate: {
+      //     path: "merchant",
+      //     populate: {
+      //       path: "user",
+      //       select: "-password",
+      //     },
+      //   },
+      // })
+      .lean();
+
+    if (!consumerCoupons?.length) {
+      return res.status(400).json({ message: "No consumer coupons found" });
+    }
+
+    // Return the consumer coupons
+    res.status(200).json(consumerCoupons);
+  } catch (error) {
+    // Handle errors
+    console.error(error);
+    res.status(500).json({ error: "Internal server error" });
+  }
+};
+
+const addConsumerCoupon = async (req, res) => {
+  try {
+    const consumerId = req.consumerId;
+
+    if (!consumerId) {
+      return res
+        .status(404)
+        .json({ error: "Consumer not found. Please log in" });
+    }
+
+    const { couponId } = req.body;
+
+    if (!couponId) {
+      return res.status(400).json({ message: "couponId is required" });
+    }
+
+    const qrCode = await generateBase64QRCode({
+      consumer: consumerId,
+      coupon: couponId,
+    });
+
```

Comment:
Before generating a qrCode, validate that the couponId is existing in the database. Otherwise throw an error.